P2PEnabledConference.js

```diff
@@ -0,0 +1,899 @@
+                this.room.myroomjid, peerJid);
+        this.peerToPeerSession.setSSRCOwnerJid(this.room.myroomjid);
+        this.peerToPeerFakeRoom = this._createFakeRoom(true);
+
+        logger.info(
+            "Created new P2P JingleSession", this.room.myroomjid, peerJid);
+
+        this.peerToPeerSession.initialize(
+            true /* initiator */, this.peerToPeerFakeRoom, this.rtc);
+
+        // NOTE one may consider to start P2P with the local tracks detached,
+        // but no data will be sent until ICE succeeds anyway. And we switch
+        // immediately once the P2P ICE connects.
+        const localTracks = this.getLocalTracks();
+
+        logger.info("Adding " + localTracks + " to P2P...");
+        this.peerToPeerSession.addLocalTracks(localTracks).then(
+            () => {
+                logger.info("Added " + localTracks + " to P2P");
+                logger.info("About to send P2P 'session-initiate'...");
+                this.peerToPeerSession.invite();
+            },
+            (error) => {
+                logger.error("Failed to add " + localTracks + " to P2P", error);
+            });
+    }
+
+    /**
+     * Method when called will decide whether it's the time to start or stop the
+     * P2P session.
+     * @param {boolean} userLeftEvent if <tt>true</tt> it means that the call
+     * originates from the user left event.
+     * @private
+     */
+    _startStopP2PSession (userLeftEvent) {
+        if (this.options.config.disableAutoP2P) {
+            logger.info("Auto P2P disabled");
+            return;
+        }
+        const peers = this.getParticipants();
+        const peerCount = peers.length;
+        const isModerator = this.isModerator();
+        // FIXME 1 peer and it must *support* P2P switching
+        const shouldBeInP2P = peerCount === 1;
+
+        logger.debug(
+            "P2P? isModerator: " + isModerator
+            + ", peerCount: " + peerCount + " => " + shouldBeInP2P);
+
+        // Clear deferred "start P2P" task
+        if (!shouldBeInP2P && this.deferredStartP2P) {
+            logger.info("Cleared deferred start P2P task");
+            window.clearTimeout(this.deferredStartP2P);
+            this.deferredStartP2P = null;
+        }
+        // Start peer to peer session
+        if (isModerator && !this.peerToPeerSession && shouldBeInP2P) {
+            const peer = peerCount && peers[0];
+
+            // Everyone is a moderator ?
+            if (isModerator && peer.getRole() === 'moderator') {
+                const myId = this.myUserId();
+                const peersId = peer.getId();
+                if (myId > peersId) {
+                    logger.debug(
+                        "Everyone's a moderator - "
+                            + "the other peer should start P2P", myId, peersId);
+                    // Abort
+                    return;
+                } else if (myId == peersId) {
+                    logger.error("The same IDs ? ", myId, peersId);
```

Comment:
Yes, it should "abort" at least, but I would not "throw", because if it's executed from Strophe packet processor for the "user joined room" presence it will remove this handler and break the whole app. It can continue without P2P though.